### PR TITLE
Add support for UniqueId selectors

### PR DESCRIPTION
**Pull request by alextu**
This should fix https://github.com/helmethair-co/scalatest-junit-runner/issues/88 by enabling discovery by UniqueIds.

For example, given a selector `[engine:scalatest]/[suite:tests.SomeTest]`, it should discover `SomeTest`.
But given `[engine:scalatest]/[suite:tests.SomeTest]/[some method]`, it will still discover `SomeTest` and execute all its tests method. 
Supporting selecting only the provided method would require some refactoring in the `Discovery` class.

### CODE REVIEW

File: src/test/java/co/helmethair/scalatest/helper/TestHelpers.java
Comment (alextu) [reply src/test/java/co/helmethair/scalatest/helper/TestHelpers.java:91]: brilliant

File: src/main/java/co/helmethair/scalatest/ScalatestEngine.java
Comment (giurim) [reply src/main/java/co/helmethair/scalatest/ScalatestEngine.java:43]: I would keep it specific, only because this way properties are accessible if needed and also has some documentation power.

File: src/main/java/co/helmethair/scalatest/ScalatestEngine.java
Comment (giurim) [inline src/main/java/co/helmethair/scalatest/ScalatestEngine.java:44]: Can we make this immutable somehow? WIth streams maybe.

File: src/main/java/co/helmethair/scalatest/ScalatestEngine.java
Comment (giurim) [reply src/main/java/co/helmethair/scalatest/ScalatestEngine.java:87]: 1: agree, I am not an expert so I would need some research what is the idiomatic way to load a class.
2: I used `Thread.currentThread().getContextClassLoader()` elsewhere and I think it should be the same as here.

File: src/main/java/co/helmethair/scalatest/ScalatestEngine.java
Comment (giurim) [reply src/main/java/co/helmethair/scalatest/ScalatestEngine.java:88]: True. I have some vague memories about some weird errors why this is here, but if tests pass without it, it should be ok.

File: src/main/java/co/helmethair/scalatest/ScalatestEngine.java
Comment (giurim) [inline src/main/java/co/helmethair/scalatest/ScalatestEngine.java:92]: Just logging an error or silently catching it could cover errors and make the build pass. It should be surfaced.

File: src/main/java/co/helmethair/scalatest/descriptor/ScalatestDescriptor.java
Comment (giurim) [inline src/main/java/co/helmethair/scalatest/descriptor/ScalatestDescriptor.java:36]: 👍

File: src/test/java/co/helmethair/scalatest/UniqueIdTest.java
Comment (giurim) [inline src/test/java/co/helmethair/scalatest/UniqueIdTest.java:17]: I think this is an important limitation can you add to the PR description?

File: src/test/java/co/helmethair/scalatest/helper/TestHelpers.java
Comment (giurim) [reply src/test/java/co/helmethair/scalatest/helper/TestHelpers.java:91]: TIL, good idea

**Review by giurim — COMMENTED** [review]
First of all, thank you for your contribution. I will dedicate some time to merge this soon.

File: src/main/java/co/helmethair/scalatest/ScalatestEngine.java
Comment (marcphilipp) [reply src/main/java/co/helmethair/scalatest/ScalatestEngine.java:87]: That's what `ReflectionSupport.loadClass` does under the hood.

File: src/main/java/co/helmethair/scalatest/ScalatestEngine.java
Comment (alextu) [reply src/main/java/co/helmethair/scalatest/ScalatestEngine.java]: Yes, I adressed it by moving the logic to the `Discover.discover` that now takes a list of class names. It should also adress all comments related to the generics since now in `ScalatestEngine` only class names are considered.

File: src/test/java/co/helmethair/scalatest/helper/TestHelpers.java
Comment (alextu) [reply src/test/java/co/helmethair/scalatest/helper/TestHelpers.java]: I removed it, thanks to @marcphilipp suggestion I used the https://junit.org/junit5/docs/current/user-guide/#testkit-engine which removes the need to create those internal details to run the tests.

File: src/main/java/co/helmethair/scalatest/ScalatestEngine.java
Comment (giurim) [reply src/main/java/co/helmethair/scalatest/ScalatestEngine.java]: Yeah, these checks are responses to some bugs.

File: src/main/java/co/helmethair/scalatest/ScalatestEngine.java
Comment (giurim) [inline src/main/java/co/helmethair/scalatest/ScalatestEngine.java:41]: Can we please change this to something like this below? No real difference just saves a few lines of code. 
```suggestion
            Stream.concat(
                discoverClassSelectors(discoveryRequest).stream(),
                discoverUniqueIdSelectors(discoveryRequest).stream()
            ).collect(Collectors.toSet())
```

**Review by giurim — COMMENTED** [review]
Nice work, I would only change one thing.

File: src/main/java/co/helmethair/scalatest/ScalatestEngine.java
Comment (alextu) [reply src/main/java/co/helmethair/scalatest/ScalatestEngine.java:41]: Nice, minor detail the returned set might be mutable `Collectors.toUnmodifiableSet()` is only available from java 10

Comment (giurim) [reply src/main/java/co/helmethair/scalatest/ScalatestEngine.java]: Thank you very much, its fine, for me the functional style matters more

**Comment by giurim** [comment]
One more thing: I cannot update your branch, main is ahead and cannot merge until it is up to date.

**Review by giurim — DISMISSED** [review]

**Comment by alextu** [comment]
> One more thing: I cannot update your branch, main is ahead and cannot merge until it is up to date.

Will do, could you also wait for @marcphilipp's approval before merging?

**Comment by giurim** [comment]
Sure thing, happy to have another set of eyes on it.

File: src/main/java/co/helmethair/scalatest/ScalatestEngine.java
Comment (marcphilipp) [inline src/main/java/co/helmethair/scalatest/ScalatestEngine.java:78]:
```suggestion
        return discoveryRequest.getSelectorsByType(ClassSelector.class).stream()
        	.map(ClassSelector::getClassName)
```

**Review by marcphilipp — DISMISSED** [review]

**Review by giurim — APPROVED** [review]

**Pull request merged**
